fdbserver/BlobManifest.actor.cpp

```diff
@@ -0,0 +1,369 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fdbclient/BackupContainer.h"
+#include "fdbserver/Knobs.h"
+#include "flow/FastRef.h"
+#include "flow/flow.h"
+#include "fdbclient/NativeAPI.actor.h"
+#include "fdbclient/BlobConnectionProvider.h"
+#include "fdbclient/FDBTypes.h"
+#include "fdbclient/KeyRangeMap.h"
+#include "fdbclient/SystemData.h"
+#include "fdbclient/BackupContainerFileSystem.h"
+#include "fdbclient/BlobGranuleReader.actor.h"
+#include "fdbserver/BlobGranuleServerCommon.actor.h"
+
+#include "flow/actorcompiler.h" // has to be last include
+#include "fmt/core.h"
+
+//
+// This module offers routines to dump or load blob manifest file, which is used for full restore from granules
+//
+
+static std::string MANIFEST_FILENAME = "manifest"; // Default manifest file name on external blob storage
+
+#define ENABLE_DEBUG_PRINT true
+template <typename... T>
+inline void dprint(fmt::format_string<T...> fmt, T&&... args) {
+	if (ENABLE_DEBUG_PRINT)
+		fmt::print(fmt, std::forward<T>(args)...);
+}
+
+// This class dumps blob manifest to external blob storage.
+class BlobManifestDumper : public ReferenceCounted<BlobManifestDumper> {
+public:
+	BlobManifestDumper(Database& db, Reference<BlobConnectionProvider> blobConn) : db_(db), blobConn_(blobConn) {}
+	virtual ~BlobManifestDumper() {}
+
+	// Execute the dumper
+	ACTOR static Future<Void> execute(Reference<BlobManifestDumper> self) {
+		try {
+			state Standalone<BlobManifest> manifest;
+			Standalone<VectorRef<KeyValueRef>> rows = wait(getSystemKeys(self));
+			manifest.rows = rows;
+			Value data = encode(manifest);
+			wait(writeToFile(self, data));
+		} catch (Error& e) {
+			dprint("WARNING: unexpected blob manifest dumper error {}\n", e.what()); // skip error handling for now
+		}
+		return Void();
+	}
+
+private:
+	// Return system keys that to be backed up
+	ACTOR static Future<Standalone<VectorRef<KeyValueRef>>> getSystemKeys(Reference<BlobManifestDumper> self) {
+		state Standalone<VectorRef<KeyValueRef>> rows;
+		loop {
+			state Transaction tr(self->db_);
+			tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			try {
+				state std::vector<KeyRangeRef> ranges = {
+					blobGranuleMappingKeys, // Map granule to workers. Track the active granules
+					blobGranuleFileKeys, // Map a granule version to granule files. Track files for a granule
+					blobGranuleHistoryKeys, // Map granule to its parents and parent bundaries. for time-travel read
+					blobRangeKeys // Key ranges managed by blob
+				};
+				for (auto range : ranges) {
+					RangeResult result = wait(tr.getRange(range, GetRangeLimits::BYTE_LIMIT_UNLIMITED));
```

Comment:
Doesn't have to be part of this PR, but you'd want to support chunking of these reads instead of doing them all at once. It'd be optimal to use a rangeStream for them, since it'd get you better performance too.

Reply:
sure. will switch getRangeStream it in next PR.